plugins/bil_pipe/bil_const_propagation.ml

```diff
@@ -0,0 +1,236 @@
+        SM.return (Bil.move v e)
+      | e ->
+        self#eval_exp e >>= fun e ->
+        self#update Top v >>= fun () ->
+        SM.return (Bil.move v e)
+
+    method eval_if cond yes no =
+      SM.get () >>= fun ctxt ->
+      self#eval_exp cond >>= fun cond ->
+      self#eval yes >>= fun yes ->
+      SM.get () >>= fun ctxt_yes ->
+      SM.put ctxt >>= fun () ->
+      self#eval no >>= fun no ->
+      SM.get () >>= fun ctxt_no ->
+      let ctxt = match has_jmps yes, has_jmps no with
+        | false, false -> meet ctxt_yes ctxt_no
+        | true, true -> ctxt
+        | false, _ -> ctxt_yes
+        | _ -> ctxt_no in
+      SM.put ctxt >>= fun () ->
+      SM.return (Bil.if_ cond yes no)
+
+    method eval_while cond body =
+      SM.get () >>= fun ctxt ->
+      SM.List.iter (defs body) ~f:(self#update Top) >>= fun () ->
+      self#eval_exp cond >>= fun cond ->
+      self#eval body >>= fun body ->
+      SM.get () >>= fun ctxt' ->
+      let ctxt = if has_jmps body then ctxt
+        else meet ctxt ctxt' in
+      SM.put ctxt >>= fun () ->
+      SM.return (Bil.while_ cond body)
+
+    method eval_jmp e : stmt r =
+      self#eval_exp e >>= fun e ->
+      SM.return (Bil.jmp e)
+  end
+end
+
+module Dead_code = struct
+
+  let is_used var bil =
+    let (||) x y = match x, y with
+      | Some r1, Some r2 -> Some (r1 || r2)
+      | Some _, None | None, Some _ -> Some true
+      | None, None -> None in
+    let is_in_free s = Set.mem (Stmt.free_vars s) var in
+    let defined = Var.equal var in
+    let rec used_in = function
+      | [] -> None
+      | s :: bil -> match s with
+        | If (_, yes, no) ->
+          let r = used_in yes || used_in no in
+          if Option.is_some r then r
+          else used_in bil
+        | While (_, body) -> used_in body || used_in bil
+        | Move (var',_) when defined var' -> Some false
+        | s when is_in_free s -> Some true
+        | _ -> used_in bil in
+    match used_in bil with
+    | None -> false
+    | Some x -> x
+
+  let eliminate bil =
+    let decr = function
+      | [] -> []
+      | _ :: scope -> scope in
+    let rec loop acc scope = function
+      | [] -> List.rev acc
+      | st :: bil ->
+        let scope = decr scope in
+        match st with
+        | Move (v, Int _) when Var.is_virtual v ->
+	  if is_used v scope then loop (st :: acc) scope bil
+	  else loop acc scope bil
+        | If (cond, yes, no) ->
+	  let yes = loop [] (yes @ scope) yes in
+	  let no  = loop [] (no @ scope) no in
+	  loop (If (cond, yes, no) :: acc) scope bil
+        | While (cond,body) ->
+          let body = loop [] (body @ bil) body in
+          loop (While (cond,body) :: acc) scope bil
+        | st -> loop (st :: acc) scope bil in
+    loop [] bil bil
+
+end
+
+module P = Propagate(Monad.State)
+
+let propagate_consts xs = Monad.State.eval ((new P.t)#eval xs) P.new_ctxt
+let eliminate_dead_code = Dead_code.eliminate
+
+let propagate_consts bil =
```

Comment:
so, you're saying that "constant propagation is constant folding, constant propagation, and dead code elimination"?